drivers/bluetooth/hci/userchan.c

```diff
@@ -0,0 +1,214 @@
+
+#define BTPROTO_HCI      1
+struct sockaddr_hci {
+	sa_family_t     hci_family;
+	unsigned short  hci_dev;
+	unsigned short  hci_channel;
+};
+#define HCI_CHANNEL_USER 1
+
+#define SOL_HCI          0
+
+#define H4_NONE          0x00
+#define H4_CMD           0x01
+#define H4_ACL           0x02
+#define H4_SCO           0x03
+#define H4_EVT           0x04
+
+static BT_STACK_NOINIT(rx_thread_stack, CONFIG_BT_RX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+
+static int uc_fd = -1;
+
+static struct net_buf *get_rx(const u8_t *buf)
+{
+	if (buf[0] == H4_EVT && (buf[1] == BT_HCI_EVT_CMD_COMPLETE ||
+				 buf[1] == BT_HCI_EVT_CMD_STATUS)) {
+		return bt_buf_get_cmd_complete(K_FOREVER);
+	}
+
+	if (buf[0] == H4_ACL) {
+		return bt_buf_get_rx(BT_BUF_ACL_IN, K_FOREVER);
+	} else {
+		return bt_buf_get_rx(BT_BUF_EVT, K_FOREVER);
+	}
+}
+
+static bool uc_ready(void)
+{
+	struct pollfd pollfd = { .fd = uc_fd, .events = POLLIN };
+
+	return (poll(&pollfd, 1, 0) == 1);
+}
+
+static void rx_thread(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	BT_DBG("started");
+
+	while (1) {
+		static u8_t frame[512];
+		struct net_buf *buf;
+		ssize_t len;
+
+		if (!uc_ready()) {
+			k_sleep(K_MSEC(20));
+			continue;
+		}
+
+		BT_DBG("calling read()");
+
+		len = read(uc_fd, frame, sizeof(frame));
+		if (len < 0) {
+			BT_ERR("Reading socket failed, errno %d", errno);
+			k_yield();
+			continue;
+		}
+
+		buf = get_rx(frame);
+		net_buf_add_mem(buf, &frame[1], len - 1);
+
+		BT_DBG("Calling bt_recv(%p)", buf);
+
+		if (frame[0] == H4_EVT && bt_hci_evt_is_prio(frame[1])) {
+			bt_recv_prio(buf);
+		} else {
+			bt_recv(buf);
+		}
+
+		k_yield();
+	}
+}
+
+static int uc_send(struct net_buf *buf)
+{
+	BT_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	if (uc_fd < 0) {
+		BT_ERR("User channel not open");
+		return -EIO;
+	}
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_OUT:
+		net_buf_push_u8(buf, H4_ACL);
+		break;
+	case BT_BUF_CMD:
+		net_buf_push_u8(buf, H4_CMD);
+		break;
+	default:
+		BT_ERR("Unknown buffer type");
+		return -EINVAL;
+	}
+
+	if (write(uc_fd, buf->data, buf->len) < 0) {
+		return -errno;
+	}
+
+	net_buf_unref(buf);
+	return 0;
+}
+
+static int user_chan_open(u16_t index)
+{
+	struct sockaddr_hci addr;
+	int fd;
+
+	fd = socket(PF_BLUETOOTH, SOCK_RAW | SOCK_CLOEXEC | SOCK_NONBLOCK,
+		    BTPROTO_HCI);
+	if (fd < 0) {
+		return -errno;
+	}
+
+	memset(&addr, 0, sizeof(addr));
+	addr.hci_family = AF_BLUETOOTH;
+	addr.hci_dev = index;
+	addr.hci_channel = HCI_CHANNEL_USER;
+
+	if (bind(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+		int err = -errno;
+
+		close(fd);
+		return err;
+	}
+
+	return fd;
+}
+
+static int uc_open(void)
+{
+	/* FIXME: Currently just hci0. Possible better options:
+	 *
+	 *  - Iterate all adapters and pick the first non-powered one
+	 *  - Use an environment variable to specify adapter
+	 *  - Use a command line option (e.g. --bt-dev hci0)
```

Comment:
I'd think the command line option would be the best option. You can add it easily already, and you can have a default one. I could add a patch for you if you'd want.

Reply:
Maybe the simplest is to use a mandatory --bt-dev command line option. Since there is really no easy choice for an adapter, force the user to specify one.

Reply:
I agree, it's better to make it mandatory in order to avoid confusion